SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialHandoverManager.cpp

```diff
@@ -0,0 +1,167 @@
+#include "SpatialView/EntityDelta.h"
+#include "SpatialView/SpatialOSWorker.h"
+#include "SpatialView/SubView.h"
+#include "SpatialView/ViewDelta.h"
+
+namespace SpatialGDK
+{
+FSpatialHandoverManager::FSpatialHandoverManager(const FSubView& InActorView, const FSubView& InPartitionView)
+	: ActorView(InActorView)
+	, PartitionView(InPartitionView)
+{
+}
+
+void FSpatialHandoverManager::Advance()
+{
+	for (const EntityDelta& Delta : PartitionView.GetViewDelta().EntityDeltas)
+	{
+		switch (Delta.Type)
+		{
+		case EntityDelta::ADD:
+			PartitionsToACK.Add(Delta.EntityId);
+			OwnedPartitions.Add(Delta.EntityId);
+			break;
+		case EntityDelta::REMOVE:
+			OwnedPartitions.Remove(Delta.EntityId);
+			break;
+		case EntityDelta::TEMPORARILY_REMOVED:
+
+			break;
+		default:
+			break;
+		}
+	}
+
+	for (const EntityDelta& Delta : ActorView.GetViewDelta().EntityDeltas)
+	{
+		switch (Delta.Type)
+		{
+		case EntityDelta::UPDATE:
+		{
+			for (const ComponentChange& Change : Delta.ComponentUpdates)
+			{
+				ApplyComponentUpdate(Delta.EntityId, Change.ComponentId, Change.Update);
+			}
+			for (const ComponentChange& Change : Delta.ComponentsRefreshed)
+			{
+				ApplyComponentRefresh(Delta.EntityId, Change.ComponentId, Change.CompleteUpdate.Data);
+			}
+			break;
+		}
+		case EntityDelta::ADD:
+			PopulateDataStore(Delta.EntityId);
+			break;
+		case EntityDelta::REMOVE:
+			DataStore.Remove(Delta.EntityId);
+			break;
+		case EntityDelta::TEMPORARILY_REMOVED:
+			DataStore.Remove(Delta.EntityId);
+			PopulateDataStore(Delta.EntityId);
+			break;
+		default:
+			break;
+		}
+	}
+}
+
+void FSpatialHandoverManager::PopulateDataStore(const Worker_EntityId EntityId)
+{
+	LBComponents2& Components = DataStore.Emplace(EntityId, LBComponents2{});
+	for (const ComponentData& Data : ActorView.GetView()[EntityId].Components)
+	{
+		switch (Data.GetComponentId())
+		{
+		case SpatialConstants::AUTHORITY_INTENTV2_COMPONENT_ID:
+			Components.Intent = AuthorityIntentV2(Data.GetUnderlying());
+			break;
+		case SpatialConstants::AUTHORITY_INTENT_ACK_COMPONENT_ID:
+			Components.IntentACK = AuthorityIntentACK(Data.GetUnderlying());
+			break;
+		default:
+			break;
+		}
+	}
+	HandleChange(EntityId, Components);
+}
+
+void FSpatialHandoverManager::HandleChange(Worker_EntityId EntityId, const LBComponents2& Components)
+{
+	if (Components.Intent.AssignmentCounter != Components.IntentACK.AssignmentCounter)
+	{
+		if (OwnedPartitions.Contains(Components.Intent.PartitionId))
+		{
+			ActorsToACK.Add(EntityId);
+		}
+		else
+		{
+			ActorsToHandover.Add(EntityId);
+		}
+	}
+}
+
+void FSpatialHandoverManager::ApplyComponentUpdate(const Worker_EntityId EntityId, const Worker_ComponentId ComponentId,
+												   Schema_ComponentUpdate* Update)
+{
+	LBComponents2& Components = DataStore[EntityId];
+	switch (ComponentId)
+	{
+	case SpatialConstants::AUTHORITY_INTENTV2_COMPONENT_ID:
+		Components.Intent.ApplyComponentUpdate(Update);
+		break;
```

Comment:
I assume we don't update the Ack component here because we assume only the auth server will have interest, so no one should see an update (except the strategy worker). Is there a chance that changes? Safer to update the Ack component here also then?

Reply:
If we want other workers to process non-auth auth-intent updates then we should put that in a subview / system for dealing with it. We shouldn't tack it onto this.

Reply:
Although I can't see why we'd want to do that.